src/Products/Viewer/Controllers/ViewerApiController.cs

```diff
@@ -430,149 +453,125 @@ private LoadDocumentEntity LoadDocument(PostedDataEntity postedData, bool loadAl
+        private MemoryStream RenderPageToMemoryStream(int pageNumberToRender, string documentGuid, string password)
         {
-            List<string> allPages = new List<string>();
-            if (globalConfiguration.Viewer.GetIsHtmlMode())
+            MemoryStream result = new MemoryStream();
+
+            var loadOptions = GetLoadOptions(password);
+
+            using (GroupDocs.Viewer.Viewer viewer = new GroupDocs.Viewer.Viewer(documentGuid, loadOptions))
             {
-                HtmlOptions htmlOptions = new HtmlOptions();
-                SetOptions(htmlOptions, password, 0);
-                // get page HTML              
-                var pages = this.GetHandler().GetPages(documentGuid, htmlOptions);
-                for (int i = 0; i < pages.Count; i++)
+                if (globalConfiguration.Viewer.GetIsHtmlMode())
                 {
-                    allPages.Add(pages[i].HtmlContent);
+                    HtmlViewOptions viewOptions =
+                        HtmlViewOptions.ForEmbeddedResources(
+                            pageNumber => result,
+                            (pageNumber, pageStream) =>
+                            {
+                                // Do not close stream as we're about to read from it
+                            });
+
+                    SetWatermarkOptions(viewOptions);
+
+                    viewer.View(viewOptions, pageNumberToRender);
                 }
-            }
-            else
-            {
-                ImageOptions imageOptions = new ImageOptions();
-                SetOptions(imageOptions, password, 0);
-                var pages = this.GetHandler().GetPages(documentGuid, imageOptions);
-                for (int i = 0; i < pages.Count; i++)
+                else
                 {
-                    byte[] bytes;
-                    using (var memoryStream = new MemoryStream())
-                    {
-                        pages[i].Stream.CopyTo(memoryStream);
-                        bytes = memoryStream.ToArray();
-                    }
-                    string encodedImage = Convert.ToBase64String(bytes);
-                    allPages.Add(encodedImage);
+                    PngViewOptions PngViewOptions = new PngViewOptions(
+                        pageNumber => result,
+                        (pageNumber, pageStream) =>
+                        {
+                            // Do not close stream as we're about to read from it
+                        });
+
+                    SetWatermarkOptions(PngViewOptions);
+
+                    viewer.View(PngViewOptions, pageNumberToRender);
                 }
             }
-            return allPages;
+
+            return result;
         }
 
-        private void SetOptions(HtmlOptions options, string password, int pageNumber)
+        private List<string> GetAllPagesContent(string password, string documentGuid, IList<Page> pages)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Make 'GetAllPagesContent' a static method.](https://app.codacy.com/gh/groupdocs-viewer/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5242614)